src/template.yml

```diff
@@ -801,15 +795,27 @@ Resources:
                 - s3:DeleteBucket
                 - s3:PutEncryptionConfiguration
                 - s3:PutBucketPolicy
-              Resource: "arn:aws:s3:::adf-shared-modules-*"
-            - Effect: Allow
-              Action: ssm:GetParameter
-              Resource:
-                - !Sub "arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/shared_modules_bucket"
-                - !Sub "arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/deployment_account_region"
+              # This must match BucketNamePrefix of the SharedModulesBucket resource
+              Resource: !Sub "adf-shared-modules-${DeploymentAccountMainRegion}-*"
       FunctionName: CrossRegionBucketHandler
       Runtime: python3.8
       Timeout: 300
+  # Created as a separate policy to avoid a circular dependency
+  CrossRegionBucketHandlerRoleSSMPolicy:
+    Type: AWS::IAM::Policy
+    Properties:
+      PolicyName: !Sub "ssm-policy-${AWS::Region}"
+      PolicyDocument:
+        Version: "2012-10-17"
+        Statement:
+          - Effect: Allow
+            Action: ssm:GetParameter
+            Resource:
+              - !Sub "arn:${AWS::Partition}:ssm:${AWS::Region}:${AWS::AccountId}:parameter/${SharedModulesBucketName}"
+              - !Sub "arn:${AWS::Partition}:ssm:${AWS::Region}:${AWS::AccountId}:parameter/deployment_account_region"
+      Roles:
+        # Created by the Serverless Transform of CrossRegionBucketHandler
+        - !GetAtt CrossRegionBucketHandlerRole
```

Comment:
Personally I don't like relying on a to-be-generated transform resource.
This makes reviewing the code very hard to follow.
If this is the only way to break-out of the circular reference, I would rather move the other policies away from the lambda function into a separate role definition as well.

Plus I think this would need to be a `!Ref` instead of a `!GetAtt`.

Reply:
I'll keep the hardcoded name for now, making it a dependency can be done in a different PR